### ISSUE TITLE
delaunayHull can reduce maxEdges

### ISSUE THREAD

**Issue by MoYummy** [recastnavigation/recastnavigation#347]
https://github.com/recastnavigation/recastnavigation/blob/05b2b8da80037887d3e79af8a59a8f0f8ed02602/Recast/Source/RecastMeshDetail.cpp#L465

For any valid triangulation, `E = 2VB + 3VI - 3`, `VB` for border vertices and `VI` for internal ones, according to https://www.math.utah.edu/~pa/MDS/triangulation.html
This can be translated to assert `nedges = npts * 3 - nhull -3`, so `maxEdges` could be reduced to save memory.

**Comment by MoYummy** [comment]
![image](https://user-images.githubusercontent.com/8344384/44647469-b8b70180-aa10-11e8-9ae7-bbcf93c54ed3.png)
![image](https://user-images.githubusercontent.com/8344384/44647432-a5a43180-aa10-11e8-8486-cd5c511c1626.png)


That must be a bug. Either edge `(11,18)` or `(13,14)` should not exist. I will take a look at what is going on.

**Comment by jakobbotsch** [comment]
Yep, that definitely does not look right. Thanks for taking a look!

**Comment by MoYummy** [comment]
![image](https://user-images.githubusercontent.com/8344384/44650467-4dbdf880-aa19-11e8-9774-e821c00e3d07.png)

After this stage, 4 edges are added, `(11,14)`, `(14,18)`, `(14,13)`, `(13,11)` and triggers new circle testing to find left face.

https://github.com/recastnavigation/recastnavigation/blob/05b2b8da80037887d3e79af8a59a8f0f8ed02602/Recast/Source/RecastMeshDetail.cpp#L405-L427
`(11,14)` finds `18` inside safe circumcircle of that with `13` and skip the `overlapEdges`.

Correct:
![image](https://user-images.githubusercontent.com/8344384/44650963-9d50f400-aa1a-11e8-8396-db4fd9b14747.png)

**Comment by MoYummy** [comment]
I have no idea what the safe circumcircle intends, but if I replace it with following, it would work.
As you can see, I am attempting to port it to Python.
~~~python
            d = vdist2(c, ptAt(pts, u))
            if d > r:
                continue
            else:
                if overlap_edges(pts, edges, nedges, s, u):
                    continue
                if overlap_edges(pts, edges, nedges, t, u):
                    continue
                
                pt = u
                c, r = circum_circle(ptAt(pts, s), ptAt(pts, t), ptAt(pts, u))
~~~

**Comment by MoYummy** [comment]
> I just reread your step by step description, and it kinda sounds like the
code runs `completeFacet()` one time too many? There's 11, 14, 13 left, so
they should just form a triangle.

You cannot judge this until you have visited them and it would be also time consuming to do the judgement, so I recommend not to do so.

I tried another solution: `tol = 1e-4`. It can also get correct result, so that will be the float number precision issue.

Imagine a circumcircle by three points and another point, what is the distance that encourages us to consider the point is inside the circumcircle and choose to update it?
Current we use `1/1000` of the radius (`tol = 1e-3`), but is that enough? What about the case where it is within `1/2000` of the radius? That question would go on endlessly. Maybe `1e-5` or `1e-6`would be better than current `1e-3` as we do the same thing as `EPS` in other places.

Here is the result of `tol = 1e-4` and it is correct.

![image](https://user-images.githubusercontent.com/8344384/44656266-d09b7f00-aa2a-11e8-9970-552048fae497.png)

![image](https://user-images.githubusercontent.com/8344384/44656184-87e3c600-aa2a-11e8-920e-7fc0e5309f8a.png)

![image](https://user-images.githubusercontent.com/8344384/44656365-28d28100-aa2b-11e8-8d8d-395785b03330.png)

**Comment by MoYummy** [comment]
I have found another issue that we do not do overlap check for first vertex.
https://github.com/recastnavigation/recastnavigation/blob/05b2b8da80037887d3e79af8a59a8f0f8ed02602/Recast/Source/RecastMeshDetail.cpp#L396-L402

![image](https://user-images.githubusercontent.com/8344384/44705141-532a4a00-aad0-11e8-875d-a31ec0ced30c.png)

For edge `(5,1)`, its first vertex is `2` and it overlaps with existing `(1,6)` (added for triangle `(1,2,6)`). Afterwards, it finds other vertices outside circumcircle with `2`.

**Comment by grahamboree** [comment]
@MoYummy Appreciate the detail you've shared digging into this.  Is there some input geometry you could provide that reproduces this issue?  Alternatively if you have a fix could you open a PR referencing this issue?
